File: _engine/csharp/v0_2_0/Request.cs
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Aifunc;

namespace Aifunc.Engine.Csharp.V0_2_0;

/// <summary>
/// HTTP transport and request/response handling for the model API.
/// Uses <see cref="HttpClient"/> — zero external dependencies.
/// </summary>
public static class Request
{
    private static readonly HttpClient SharedClient = CreateClient();

    private static readonly Regex Fence = new(
        @"^```(?:json)?\s*\n?(.*?)\n?```$",
        RegexOptions.Singleline | RegexOptions.Compiled);

    private static HttpClient CreateClient()
    {
        var client = new HttpClient();
        client.Timeout = Timeout.InfiniteTimeSpan;
        return client;
    }

    // -------------------------------------------------------------------------
    // Request builders
    // -------------------------------------------------------------------------

    /// <summary>Builds request params for a standard (non-streaming) call.</summary>
    public static Types.ModelRequestParams BuildRequest(
        Types.AIFuncArtifact artifact, string prompt, AIFuncConfig config)
    {
        if (string.IsNullOrWhiteSpace(config.Model))
            throw new AIFuncException(
                "AIFuncConfig.model is required when mock mode is disabled");

        var messages = new List<Dictionary<string, string>>
        {
            new() { ["role"] = "user", ["content"] = prompt },
        };
        var responseFormat = new Dictionary<string, string> { ["type"] = "json_object" };
        var resolved = ResolveModelParams(artifact, config.Model!);

        return new Types.ModelRequestParams(
            config.Model!,
            messages,
            config.Temperature ?? resolved.Temperature ?? LegacyTemperature(artifact),
            config.TopP        ?? resolved.TopP,
            config.MaxTokens   ?? resolved.MaxTokens   ?? LegacyMaxTokens(artifact),
            responseFormat,
            stream: false);
    }

    /// <summary>
    /// Builds request params for a streaming SSE call.
    /// Sets <c>stream: true</c> and omits <c>response_format</c>.
    /// </summary>
    public static Types.ModelRequestParams BuildStreamRequest(
        Types.AIFuncArtifact artifact, string prompt, AIFuncConfig config)
    {
        if (string.IsNullOrWhiteSpace(config.Model))
            throw new AIFuncException(
                "AIFuncConfig.model is required when mock mode is disabled");

        var messages = new List<Dictionary<string, string>>
        {
            new() { ["role"] = "user", ["content"] = prompt },
        };
        var resolved = ResolveModelParams(artifact, config.Model!);

        return new Types.ModelRequestParams(
            config.Model!,
            messages,
            config.Temperature ?? resolved.Temperature ?? LegacyTemperature(artifact),
            config.TopP        ?? resolved.TopP,
            config.MaxTokens   ?? resolved.MaxTokens   ?? LegacyMaxTokens(artifact),
            responseFormat: null,
            stream: true);
    }

    // -------------------------------------------------------------------------
    // Standard (non-streaming) send
    // -------------------------------------------------------------------------

    /// <summary>Sends a standard request to the model API asynchronously.</summary>
    public static async Task<Dictionary<string, object?>> SendRequestAsync(
        AIFuncConfig config, Types.ModelRequestParams parameters, int timeoutMs)
    {
        var endpoint = ResolveEndpoint(config);
        var bodyJson = Json.Stringify(parameters.ToJsonMap());

        using var cts     = new CancellationTokenSource(TimeSpan.FromMilliseconds(timeoutMs));
        using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Content = new StringContent(bodyJson, Encoding.UTF8, "application/json");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", config.ApiKey);

        HttpResponseMessage response;
        try
        {
            response = await SharedClient.SendAsync(request, cts.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException ex)
        {
            throw new AIFuncException("Model API request timed out after " + timeoutMs + "ms", ex);
        }
        catch (HttpRequestException ex)
        {
            throw new AIFuncException("Model API request failed: " + ex.Message, ex);
        }

        using (response)
        {
            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            if ((int)response.StatusCode >= 400)
            {
                var preview = body.Length > 500 ? body.Substring(0, 500) : body;
                throw new AIFuncException(
                    "Model API returned " + (int)response.StatusCode + ": " + preview);
            }
            return Json.ParseObject(body);
        }
    }

    // -------------------------------------------------------------------------
    // Streaming SSE send
    // -------------------------------------------------------------------------

    /// <summary>
    /// Opens an SSE stream to the model API and yields text tokens as they arrive.
    /// Cancellation is handled via <paramref name="cancellationToken"/>.
    /// </summary>
    public static async IAsyncEnumerable<string> SendStreamRequestAsync(
        AIFuncConfig config,
        Types.ModelRequestParams parameters,
        int timeoutMs,
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var endpoint = ResolveEndpoint(config);
        var bodyJson = Json.Stringify(parameters.ToJsonMap());

        using var timeoutCts = new CancellationTokenSource(TimeSpan.FromMilliseconds(timeoutMs));
        using var linkedCts  = CancellationTokenSource.CreateLinkedTokenSource(
            timeoutCts.Token, cancellationToken);

        using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Content = new StringContent(bodyJson, Encoding.UTF8, "application/json");
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", config.ApiKey);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("text/event-stream"));
        request.Headers.CacheControl = new CacheControlHeaderValue { NoCache = true };

        HttpResponseMessage response;
        try
        {
            response = await SharedClient
                .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, linkedCts.Token)
                .ConfigureAwait(false);
        }
        catch (OperationCanceledException ex) when (timeoutCts.IsCancellationRequested)
        {
            throw new AIFuncException(
                "Model API stream request timed out after " + timeoutMs + "ms", ex);
        }
        catch (OperationCanceledException) { yield break; }
        catch (HttpRequestException ex)
        {
            throw new AIFuncException("Model API request failed: " + ex.Message, ex);
        }

        using (response)
        {
            if ((int)response.StatusCode >= 400)
            {
                var errBody = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
                var preview = errBody.Length > 500 ? errBody.Substring(0, 500) : errBody;
                throw new AIFuncException(
                    "Model API returned " + (int)response.StatusCode + ": " + preview);
            }

            await using var stream = await response.Content
                .ReadAsStreamAsync().ConfigureAwait(false);
            using var reader = new StreamReader(stream, Encoding.UTF8);

            while (!reader.EndOfStream && !linkedCts.Token.IsCancellationRequested)
            {
                string? line;
                try
                {
                    line = await reader.ReadLineAsync().ConfigureAwait(false);
                }
                catch (OperationCanceledException) { yield break; }
                catch (IOException)               { yield break; }

                if (line is null) break;
                if (!line.StartsWith("data:", StringComparison.Ordinal)) continue;

                var payload = line.Substring(5).Trim();
                if (payload == "[DONE]") yield break;

                var content = ParseSseDelta(payload);
                if (content is not null && content.Length > 0)
                    yield return content;
            }
        }
    }

    // -------------------------------------------------------------------------
    // Response parsing (non-streaming)
    // -------------------------------------------------------------------------

    /// <summary>Parses a non-streaming model response map into the output dictionary.</summary>
    public static Dictionary<string, object?> ParseResponse(Dictionary<string, object?> responseMap)
    {
        if (!responseMap.TryGetValue("choices", out var choicesVal)
            || choicesVal is not IList choices || choices.Count == 0)
        {
            throw new AIFuncException("Model response contains no choices");
        }
        if (AsDict(choices[0]) is not { } firstMap)
            throw new AIFuncException("Model response choice is not an object");
        if (!firstMap.TryGetValue("message", out var msgVal) || AsDict(msgVal) is not { } msgMap)
            throw new AIFuncException("Model response message is not an object");

        msgMap.TryGetValue("content", out var contentVal);
        var content = contentVal is string cs ? cs.Trim() : "";
        var fence = Fence.Match(content);
        if (fence.Success) content = fence.Groups[1].Value.Trim();

        try
        {
            var parsed = Json.Parse(content);
            if (parsed is not Dictionary<string, object?> result)
                throw new AIFuncException("Model output is not a JSON object");
            return result;
        }
        catch (Json.JsonParseException e)
        {
            var preview = content.Length > 200 ? content.Substring(0, 200) : content;
            throw new AIFuncException("Failed to parse model output as JSON: " + preview, e);
        }
    }

    // -------------------------------------------------------------------------
    // SSE delta parser
    // -------------------------------------------------------------------------

    private static string? ParseSseDelta(string payload)
    {
        try
        {
            var parsed = Json.Parse(payload);
            if (parsed is not Dictionary<string, object?> obj) return null;
            if (!obj.TryGetValue("choices", out var cv) || cv is not IList choices || choices.Count == 0)
                return null;
            if (AsDict(choices[0]) is not { } first) return null;
            if (!first.TryGetValue("delta", out var dv) || AsDict(dv) is not { } delta) return null;
            if (!delta.TryGetValue("content", out var content) || content is not string s) return null;
            return s;
        }
        catch (Json.JsonParseException) { return null; }
    }

    // -------------------------------------------------------------------------
    // Model param resolution
    // -------------------------------------------------------------------------

    private sealed class StandardParams
    {
        public double? Temperature { get; }
        public double? TopP { get; }
        public int? MaxTokens { get; }

        public StandardParams(double? temperature, double? topP, int? maxTokens)
        {
            Temperature = temperature;
            TopP        = topP;
            MaxTokens   = maxTokens;
        }
    }

    private static StandardParams ResolveModelParams(Types.AIFuncArtifact artifact, string model)
    {
        var mp = artifact.ModelParams;
        if (mp is null) return new StandardParams(null, null, null);
        if (!mp.TryGetValue("rules", out var rulesVal) || rulesVal is not IList rules)
            return new StandardParams(null, null, null);

        foreach (var ruleObj in rules)
        {
            if (AsDict(ruleObj) is not { } rule) continue;
            if (!rule.TryGetValue("match", out var matchVal) || AsDict(matchVal) is not { } match)
                continue;
            if (!MatchesRule(match, model)) continue;
            if (!rule.TryGetValue("params", out var paramsVal) || AsDict(paramsVal) is not { } parameters)
                continue;

            parameters.TryGetValue("temperature", out var temp);
            parameters.TryGetValue("topP",        out var topP);
            parameters.TryGetValue("maxTokens",   out var maxTokens);
            return new StandardParams(ToDouble(temp), ToDouble(topP), ToInt(maxTokens));
        }
        return new StandardParams(null, null, null);
    }

    private static bool MatchesRule(Dictionary<string, object?> match, string model)
    {
        match.TryGetValue("model",   out var mModel);
        match.TryGetValue("models",  out var mModels);
        match.TryGetValue("pattern", out var mPattern);

        if (mModel is string ms && ms == model) return true;
        if (mModels is IList modelsList)
            foreach (var v in modelsList)
                if (model.Equals(v)) return true;
        if (mPattern is string pattern && !string.IsNullOrWhiteSpace(pattern))
            return GlobMatch(pattern, model);

        var noModel   = mModel   is not string mStr || string.IsNullOrWhiteSpace(mStr);
        var noModels  = mModels  is not IList  ml   || ml.Count == 0;
        var noPattern = mPattern is not string pStr || string.IsNullOrWhiteSpace(pStr);
        return noModel && noModels && noPattern;
    }

    private static bool GlobMatch(string pattern, string value)
    {
        var regex = "^" + Regex.Escape(pattern).Replace("\\*", ".*") + "$";
        return Regex.IsMatch(value, regex);
    }

    private static double? LegacyTemperature(Types.AIFuncArtifact artifact) =>
        ToDouble(artifact.Model is null ? null
            : artifact.Model.TryGetValue("temperature", out var t) ? t : null);

    private static int? LegacyMaxTokens(Types.AIFuncArtifact artifact) =>
        ToInt(artifact.Model is null ? null
            : artifact.Model.TryGetValue("maxTokens", out var t) ? t : null);

    private static string ResolveEndpoint(AIFuncConfig config)
    {
        if (string.IsNullOrWhiteSpace(config.BaseUrl))
            throw new AIFuncException(
                "AIFuncConfig.baseUrl is required when mock mode is disabled");
        if (string.IsNullOrWhiteSpace(config.ApiKey))
            throw new AIFuncException(
                "AIFuncConfig.apiKey is required when mock mode is disabled");

        var base_ = config.BaseUrl!.TrimEnd();
        if (base_.EndsWith("/")) base_ = base_.Substring(0, base_.Length - 1);
        return base_.EndsWith("/chat/completions", StringComparison.Ordinal)
            ? base_
            : base_ + "/chat/completions";
    }

    private static double? ToDouble(object? v)
    {
        if (v is null) return null;
        if (v is byte or sbyte or short or ushort or int or uint or long or ulong
               or float or double or decimal)
            return Convert.ToDouble(v);
        return null;
    }

    private static int? ToInt(object? v)
    {
        if (v is null) return null;
        if (v is byte or sbyte or short or ushort or int or uint or long or ulong
               or float or double or decimal)
            return Convert.ToInt32(v);
        return null;
    }

    private static Dictionary<string, object?>? AsDict(object? v)
    {
        if (v is Dictionary<string, object?> d) return d;
        if (v is IDictionary raw)
        {
            var c = new Dictionary<string, object?>();
            foreach (DictionaryEntry e in raw)
                c[Convert.ToString(e.Key) ?? ""] = e.Value;
            return c;
        }
        return null;
    }
}

File: _engine/java/v0_2_0/Request.java
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package aifunc._engine.java.v0_2_0;

import aifunc.AIFuncException;
import aifunc.AIFuncConfig;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * HTTP transport and request/response handling for the model API.
 *
 * <p>Uses {@link java.net.http.HttpClient} (Java 11+) — zero external dependencies.
 */
public final class Request {

    private Request() {}

    private static final HttpClient SHARED_CLIENT = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(30))
            .build();

    private static final Pattern FENCE = Pattern.compile(
            "(?s)^```(?:json)?\\s*\\n?(.*?)\\n?```$");

    // -------------------------------------------------------------------------
    // Request builders
    // -------------------------------------------------------------------------

    /** Builds params for a standard (non-streaming) call. */
    public static Types.ModelRequestParams buildRequest(
            Types.AIFuncArtifact artifact, String prompt, AIFuncConfig config) {

        if (config.getModel() == null || config.getModel().isBlank()) {
            throw new AIFuncException(
                    "AIFuncConfig.model is required when mock mode is disabled");
        }

        List<Map<String, String>> messages = buildMessages(prompt);
        Map<String, String> responseFormat = new LinkedHashMap<>();
        responseFormat.put("type", "json_object");

        StandardParams resolved = resolveModelParams(artifact, config.getModel());
        Double  temperature = coalesceDouble(config.getTemperature(), resolved.temperature, legacyTemperature(artifact));
        Double  topP        = coalesceDouble(config.getTopP(),         resolved.topP,        null);
        Integer maxTokens   = coalesceInt(config.getMaxTokens(),   resolved.maxTokens,   legacyMaxTokens(artifact));

        return new Types.ModelRequestParams(
                config.getModel(), messages, temperature, topP, maxTokens, responseFormat, false);
    }

    /**
     * Builds params for a streaming call.
     * Sets {@code stream: true} and omits {@code response_format}.
     */
    public static Types.ModelRequestParams buildStreamRequest(
            Types.AIFuncArtifact artifact, String prompt, AIFuncConfig config) {

        if (config.getModel() == null || config.getModel().isBlank()) {
            throw new AIFuncException(
                    "AIFuncConfig.model is required when mock mode is disabled");
        }

        List<Map<String, String>> messages = buildMessages(prompt);
        StandardParams resolved = resolveModelParams(artifact, config.getModel());
        Double  temperature = coalesceDouble(config.getTemperature(), resolved.temperature, legacyTemperature(artifact));
        Double  topP        = coalesceDouble(config.getTopP(),         resolved.topP,        null);
        Integer maxTokens   = coalesceInt(config.getMaxTokens(),   resolved.maxTokens,   legacyMaxTokens(artifact));

        return new Types.ModelRequestParams(
                config.getModel(), messages, temperature, topP, maxTokens, null, true);
    }

    // -------------------------------------------------------------------------
    // Async send — standard
    // -------------------------------------------------------------------------

    public static CompletableFuture<Map<String, Object>> sendRequestAsync(
            AIFuncConfig config, Types.ModelRequestParams params, int timeoutMs) {

        if (config.getBaseUrl() == null || config.getBaseUrl().isBlank()) {
            return failedFuture(new AIFuncException(
                    "AIFuncConfig.baseUrl is required when mock mode is disabled"));
        }
        if (config.getApiKey() == null || config.getApiKey().isBlank()) {
            return failedFuture(new AIFuncException(
                    "AIFuncConfig.apiKey is required when mock mode is disabled"));
        }

        String endpoint = resolveEndpoint(config.getBaseUrl());
        String bodyJson = Json.stringify(params.toJsonMap());

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .timeout(Duration.ofMillis(timeoutMs))
                .header("Content-Type",  "application/json")
                .header("Authorization", "Bearer " + config.getApiKey())
                .POST(HttpRequest.BodyPublishers.ofString(bodyJson))
                .build();

        return SHARED_CLIENT.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                .thenApply(response -> {
                    if (response.statusCode() >= 400) {
                        String preview = response.body();
                        if (preview.length() > 500) preview = preview.substring(0, 500);
                        throw new AIFuncException(
                                "Model API returned " + response.statusCode() + ": " + preview);
                    }
                    return Json.parseObject(response.body());
                });
    }

    // -------------------------------------------------------------------------
    // Async send — streaming (SSE)
    // -------------------------------------------------------------------------

    /**
     * Opens an SSE stream to the model API and delivers text tokens to
     * {@code tokenConsumer} as they arrive. Runs on a background thread.
     *
     * <p>Cancellation is signalled by setting {@code cancelled} to {@code true};
     * the SSE read loop checks this flag before each line.
     *
     * @return a future that completes when the stream ends (normally or via
     *         cancellation), or fails with {@link AIFuncException} on error
     */
    public static CompletableFuture<Void> sendStreamRequestAsync(
            AIFuncConfig config,
            Types.ModelRequestParams params,
            int timeoutMs,
            Consumer<String> tokenConsumer,
            AtomicBoolean cancelled) {

        if (config.getBaseUrl() == null || config.getBaseUrl().isBlank()) {
            return failedFuture(new AIFuncException(
                    "AIFuncConfig.baseUrl is required when mock mode is disabled"));
        }
        if (config.getApiKey() == null || config.getApiKey().isBlank()) {
            return failedFuture(new AIFuncException(
                    "AIFuncConfig.apiKey is required when mock mode is disabled"));
        }

        String endpoint = resolveEndpoint(config.getBaseUrl());
        String bodyJson = Json.stringify(params.toJsonMap());

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .timeout(Duration.ofMillis(timeoutMs))
                .header("Content-Type",  "application/json")
                .header("Authorization", "Bearer " + config.getApiKey())
                .header("Accept",        "text/event-stream")
                .header("Cache-Control", "no-cache")
                .POST(HttpRequest.BodyPublishers.ofString(bodyJson))
                .build();

        return SHARED_CLIENT
                .sendAsync(request, HttpResponse.BodyHandlers.ofInputStream())
                .thenApply(response -> {
                    if (response.statusCode() >= 400) {
                        try (java.io.InputStream body = response.body();
                             BufferedReader br = new BufferedReader(
                                     new InputStreamReader(body))) {
                            StringBuilder sb = new StringBuilder();
                            String line;
                            while ((line = br.readLine()) != null
                                    && sb.length() < 500) {
                                sb.append(line);
                            }
                            throw new AIFuncException(
                                    "Model API returned "
                                    + response.statusCode() + ": " + sb);
                        } catch (java.io.IOException e) {
                            throw new AIFuncException(
                                    "Model API returned " + response.statusCode());
                        }
                    }
                    readSSEStream(response.body(), tokenConsumer, cancelled);
                    return (Void) null;
                });
    }

    // -------------------------------------------------------------------------
    // SSE reader
    // -------------------------------------------------------------------------

    static void readSSEStream(
            java.io.InputStream inputStream,
            Consumer<String> tokenConsumer,
            AtomicBoolean cancelled) {

        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(inputStream,
                        java.nio.charset.StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (cancelled != null && cancelled.get()) return;
                if (!line.startsWith("data:")) continue;

                String payload = line.substring(5).strip();
                if ("[DONE]".equals(payload)) return;

                try {
                    Object parsed = Json.parse(payload);
                    if (!(parsed instanceof Map)) continue;
                    @SuppressWarnings("unchecked")
                    Map<String, Object> obj = (Map<String, Object>) parsed;

                    Object choices = obj.get("choices");
                    if (!(choices instanceof List) || ((List<?>) choices).isEmpty()) continue;
                    Object first = ((List<?>) choices).get(0);
                    if (!(first instanceof Map)) continue;

                    @SuppressWarnings("unchecked")
                    Object delta = ((Map<String, Object>) first).get("delta");
                    if (!(delta instanceof Map)) continue;

                    @SuppressWarnings("unchecked")
                    Object content = ((Map<String, Object>) delta).get("content");
                    if (content instanceof String && !((String) content).isEmpty()) {
                        tokenConsumer.accept((String) content);
                    }
                } catch (Json.JsonParseException ignored) {
                    // skip malformed SSE lines
                }
            }
        } catch (java.io.IOException ignored) {
            // connection closed — normal end of stream
        }
    }

    // -------------------------------------------------------------------------
    // Response parsing
    // -------------------------------------------------------------------------

    @SuppressWarnings("unchecked")
    public static Map<String, Object> parseResponse(Map<String, Object> responseMap) {
        Object choicesVal = responseMap.get("choices");
        if (!(choicesVal instanceof List) || ((List<?>) choicesVal).isEmpty()) {
            throw new AIFuncException("Model response contains no choices");
        }
        Object first = ((List<?>) choicesVal).get(0);
        if (!(first instanceof Map)) {
            throw new AIFuncException("Model response choice is not an object");
        }
        Object msgVal = ((Map<?, ?>) first).get("message");
        if (!(msgVal instanceof Map)) {
            throw new AIFuncException("Model response message is not an object");
        }
        Object contentVal = ((Map<?, ?>) msgVal).get("content");
        String content = (contentVal instanceof String) ? ((String) contentVal).strip() : "";

        Matcher fence = FENCE.matcher(content);
        if (fence.matches()) content = fence.group(1).strip();

        try {
            Object parsed = Json.parse(content);
            if (!(parsed instanceof Map)) {
                throw new AIFuncException("Model output is not a JSON object");
            }
            return (Map<String, Object>) parsed;
        } catch (Json.JsonParseException e) {
            String preview = content.length() > 200 ? content.substring(0, 200) : content;
            throw new AIFuncException(
                    "Failed to parse model output as JSON: " + preview, e);
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    static <T> CompletableFuture<T> failedFuture(Throwable ex) {
        CompletableFuture<T> future = new CompletableFuture<>();
        future.completeExceptionally(ex);
        return future;
    }

    private static String resolveEndpoint(String baseUrl) {
        String base = baseUrl.stripTrailing();
        if (base.endsWith("/")) base = base.substring(0, base.length() - 1);
        return base.endsWith("/chat/completions") ? base : base + "/chat/completions";
    }

    private static List<Map<String, String>> buildMessages(String prompt) {
        List<Map<String, String>> messages = new ArrayList<>();
        Map<String, String> msg = new LinkedHashMap<>();
        msg.put("role",    "user");
        msg.put("content", prompt);
        messages.add(msg);
        return messages;
    }

    private static Double coalesceDouble(Double... values) {
        for (Double v : values) if (v != null) return v;
        return null;
    }

    private static Integer coalesceInt(Integer... values) {
        for (Integer v : values) if (v != null) return v;
        return null;
    }

    // -------------------------------------------------------------------------
    // Model param resolution
    // -------------------------------------------------------------------------

    private static final class StandardParams {
        final Double  temperature;
        final Double  topP;
        final Integer maxTokens;

        StandardParams(Double t, Double p, Integer m) {
            this.temperature = t; this.topP = p; this.maxTokens = m;
        }
    }

    @SuppressWarnings("unchecked")
    private static StandardParams resolveModelParams(
            Types.AIFuncArtifact artifact, String model) {
        Map<String, Object> mp = artifact.getModelParams();
        if (mp == null) return new StandardParams(null, null, null);

        Object rulesVal = mp.get("rules");
        if (!(rulesVal instanceof List)) return new StandardParams(null, null, null);

        for (Object ruleObj : (List<?>) rulesVal) {
            if (!(ruleObj instanceof Map)) continue;
            Map<String, Object> rule = (Map<String, Object>) ruleObj;
            Object matchVal = rule.get("match");
            if (!(matchVal instanceof Map)) continue;
            if (!matchesRule((Map<String, Object>) matchVal, model)) continue;
            Object paramsVal = rule.get("params");
            if (!(paramsVal instanceof Map)) continue;
            Map<String, Object> params = (Map<String, Object>) paramsVal;
            return new StandardParams(toDouble(params.get("temperature")),
                    toDouble(params.get("topP")), toInt(params.get("maxTokens")));
        }
        return new StandardParams(null, null, null);
    }

    private static boolean matchesRule(Map<String, Object> match, String model) {
        Object mModel = match.get("model");
        if (mModel instanceof String && ((String) mModel).equals(model)) return true;
        Object mModels = match.get("models");
        if (mModels instanceof List) {
            for (Object v : (List<?>) mModels) if (model.equals(v)) return true;
        }
        Object mPattern = match.get("pattern");
        if (mPattern instanceof String) return globMatch((String) mPattern, model);

        boolean noModel   = !(mModel   instanceof String) || ((String) mModel).isBlank();
        boolean noModels  = !(mModels  instanceof List)   || ((List<?>) mModels).isEmpty();
        boolean noPattern = !(mPattern instanceof String) || ((String) mPattern).isBlank();
        return noModel && noModels && noPattern;
    }

    private static boolean globMatch(String pattern, String value) {
        return value.matches("^" + Pattern.quote(pattern).replace("\\*", ".*") + "$");
    }

    @SuppressWarnings("unchecked")
    private static Double legacyTemperature(Types.AIFuncArtifact artifact) {
        Map<String, Object> m = artifact.getModel();
        return (m != null) ? toDouble(m.get("temperature")) : null;
    }

    @SuppressWarnings("unchecked")
    private static Integer legacyMaxTokens(Types.AIFuncArtifact artifact) {
        Map<String, Object> m = artifact.getModel();
        return (m != null) ? toInt(m.get("maxTokens")) : null;
    }

    private static Double toDouble(Object v) {
        return (v instanceof Number) ? ((Number) v).doubleValue() : null;
    }

    private static Integer toInt(Object v) {
        return (v instanceof Number) ? ((Number) v).intValue() : null;
    }
}
